File: Ch_3_Code/Monoid.hs
-- (<>) infix ki tarah kaam krega
import Data.Monoid (mempty, (<>))

main :: IO ()
main = do
    let str1 = "I love "
    let str2 = "Her"
    let str3 = "jannat aur tu dono nasib nhi hue"

-- issi tararh ++ operator kaam krta hai in haskell
-- but why java does not have this operator.......because + operatoer is overloadded in java
    let result = str1 <> str2 <> str3
    putStrLn $ "Result of str1 <> str2 <> str3: " ++ result

    -- Demonstrating the use of mempty
    let empty = mempty :: String
    putStrLn $ "Result of str1 <> mempty: " ++ (str1 <> empty)

File: Ch_3_Code/Monoid.java
public interface Monoid<T> {
    T mempty();
    T mappend(T x, T y);
}

public class StringMonoid implements Monoid<String> {
    @Override
    public String mempty() {
        return "";
    }

    @Override
    public String mappend(String x, String y) {
        return x + y;
    }
}

public class Monoid {
    public static void main(String[] args) {
        Monoid<String> stringMonoid = new StringMonoid();
        
        String empty = stringMonoid.mempty();
        String result = stringMonoid.mappend("kyuin", "pta");
        
        System.out.println("khali dil: " + empty);
        System.out.println("muqambal: " + result);
    }
}
